src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java

```diff
@@ -366,6 +369,87 @@ String readErrorMessage(InputStream in) throws IOException {
         return message.toString();
     }
 
+    void processCompletionStatus(IOException ioe, String cmd, InputStream sis) throws AgentLoadException, IOException {
+        // Read the command completion status
+        int completionStatus;
+        try {
+            completionStatus = readInt(sis);
+        } catch (IOException x) {
+            sis.close();
+            if (ioe != null) {
+                throw ioe;
+            } else {
+                throw x;
+            }
+        }
+        if (completionStatus != 0) {
+            // read from the stream and use that as the error message
+            String message = readErrorMessage(sis);
+            sis.close();
+
+            // In the event of a protocol mismatch then the target VM
+            // returns a known error so that we can throw a reasonable
+            // error.
+            if (completionStatus == ATTACH_ERROR_BADVERSION) {
+                throw new IOException("Protocol mismatch with target VM");
+            }
+
+            // Special-case the "load" command so that the right exception is
+            // thrown.
+            if (cmd.equals("load")) {
+                String msg = "Failed to load agent library";
+                if (!message.isEmpty()) {
+                    msg += ": " + message;
+                }
+                throw new AgentLoadException(msg);
+            } else {
+                if (message.isEmpty()) {
+                    message = "Command failed in target VM";
+                }
+                throw new AttachOperationFailedException(message);
+            }
+        }
+    }
+
+    /*
+     * InputStream for the socket connection to get target VM
+     */
+    static abstract class SocketInputStream extends InputStream {
+        private final long fd;
+
+        public SocketInputStream(long fd) {
+            this.fd = fd;
+        }
+
+        protected abstract int readImpl(long fd, byte[] bs, int off, int len) throws IOException;
+        protected abstract void closeImpl(long fd) throws IOException;
```

Comment:
If the subclasses all override these in exactly the same way then these do not need to be abstract and can simply delegate to VirtualMachineImpl.xxx

Reply:
All Posix OS platforms call the same VirtualMachineImpl.read/write/etc, but Windows is an unusual guy, it calls VirtualMachineImpl.readPipe/writePipe/etc

Reply:
That is a shame, though perhaps we could just rename those methods on Windows?

This also raises the obvious question can we in fact reduce this to just a Windows and Posix version, or are there other differences between Linux, macOS etc that have to be accounted for? Though in that case we could introduce a shared superclass for the Posix platforms. Next steps perhaps?

Reply:
The remaining code for the POSIX system is exactly the same, except that Linux will make an extra attempt to find java_pid in cgroup environment.